src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs

```diff
@@ -24,8 +24,12 @@ internal sealed class TagHelperBinder
     /// <param name="descriptors">The descriptors that the <see cref="TagHelperBinder"/> will pull from.</param>
     public TagHelperBinder(string tagHelperPrefix, IEnumerable<TagHelperDescriptor> descriptors)
     {
+        // To prevent creating a dictionary that gets resized many times, we attempt to guess at a
+        // suitable capacity.
+        var descriptorCount = (descriptors is ICollection<TagHelperDescriptor> bounded) ? bounded.Count : descriptors.Count();
```

Comment:
I'll let the compiler team comment since they own the code, but I would probably just update the `descriptors` parameter to be `IReadOnlyList<TagHelperDescriptor>`, and roll that through wherever it needs to go in the codebase. This is an internal API, so I would be surprised if the change affected any public APIs.

From a quick look, I wouldn't expect the `.Count()` to actually be called outside of test code anyway, so maybe it doesn't matter. The only non-test code that calls this seems to always pass an array.

Reply:
If you end up changing this, I would recommend changing to the underlying type (List<ItagHelperDescriptor>?) to avoid boxing the struct enumerator when you foreach.

Reply:
+💯

Reply:
@DustinCampbell, would this change interfere with any of your existing refactorings in this area?

Reply:
Nope.

Reply:
Ok, so I looked through the call stack here, and changing the type to something more concrete than `IReadOnlyList` would require breaking the public API (or adding an internal API version, preferably returning `ImmutableArray`).  This is something we can do in the 17.7 timeframe, but to avoid dragging this PR on, I'd suggest simply going to `IReadOnlyList` for now and filing a followup issue to move this to ImmutableArray at a later point.

Reply:
Agreed. I took a look and came to the same conclusion. This change is better than doing nothing.

Reply:
@333fred, I changed to use `IReadOnlyList` as requested. It touches several files, but doesn't seem too scary.